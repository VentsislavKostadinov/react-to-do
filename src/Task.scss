@import './font.scss';

    .list-group-horizontal {
        font-family: "Poppins-Regular", sans-serif;
        background: #f4f4f4;
        margin-top: 2rem !important;
        align-items: center !important;
        justify-content: center !important;
        text-align: center !important;

        .list-group-item {
            border: none;
            background: inherit;
        }

        &:hover {
            background: lightgoldenrodyellow;
        }

    }
