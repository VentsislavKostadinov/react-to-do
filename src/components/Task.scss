@import '../style/font.scss';
@import '../style/variables';
@import '../style/mixins';

.task {
    @include task-styling;
    align-items: baseline;

    .task-description {
        @include overflow-text-description;
    }
    .btn {
        color: inherit;
    }
}

@media only screen and (max-width: 576px) {
    .container {
        padding-left: 0.5rem !important;
        padding-right: 0.5rem !important;
    }
}
