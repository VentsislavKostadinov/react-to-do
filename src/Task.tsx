import React, { useState } from 'react'
import { TaskProps } from './model/TaskProps.types'
import { CommonInput } from './common/CommonInput'
import { CommonButton } from './common/CommonButton'
import { ListGroup, Form } from 'react-bootstrap'
import './Task.scss'

export const Task = ({ task, onEdit, onDelete, onComplete }: TaskProps) => {
    const [isEditing, setIsEditing] = useState<boolean>(false)
    const [editedTask, setEditedTask] = useState<
        TaskProps['task']['description']
    >(task.description)
    const [validated, setValidated] = useState<boolean>(false)

    const handleSave = (event: React.FormEvent) => {
        event.preventDefault()

        if (editedTask === '') {
            setValidated(true)
        } else {
            onEdit(task.id, editedTask)
            setIsEditing(false)
        }
    }
    const saveEditedTask = (e: React.ChangeEvent<HTMLInputElement>) => {
        setEditedTask(e.target.value)
    }
    const handleEditTask = () => {
        setIsEditing(true)
    }

    const handleOnComplete = () => {
        onComplete(task.id)
    }

    const handleDelete = () => {
        onDelete(task.id)
    }

    return (
        <ListGroup horizontal className="task">
            <ListGroup.Item>
                <CommonInput
                    type="checkbox"
                    className="input-checkbox"
                    data-testid="input-checkbox"
                    checked={task.completed}
                    handleChange={handleOnComplete}
                ></CommonInput>
            </ListGroup.Item>
            {isEditing ? (
                <ListGroup.Item>
                    <Form
                        className="d-flex"
                        noValidate
                        validated={validated}
                        onSubmit={handleSave}
                    >
                        <CommonInput
                            className="input-text-edit"
                            data-testid="input-text-edit"
                            type="text"
                            value={editedTask}
                            checked={task.completed}
                            handleChange={saveEditedTask}
                        />
                        <CommonButton
                            text="Save"
                            variant="transparent"
                            type="submit"
                        />
                    </Form>
                </ListGroup.Item>
            ) : (
                <>
                    <ListGroup.Item>{task.description}</ListGroup.Item>
                    <ListGroup.Item>
                        <CommonButton
                            text="Edit"
                            variant="transparent"
                            type="button"
                            handleClick={handleEditTask}
                        />
                    </ListGroup.Item>
                </>
            )}
            <ListGroup.Item>
                <CommonButton
                    text="Delete"
                    variant="transparent"
                    type="button"
                    handleClick={handleDelete}
                />
            </ListGroup.Item>
        </ListGroup>
    )
}
