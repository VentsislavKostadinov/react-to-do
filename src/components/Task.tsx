import React, { useState } from 'react'
import { TaskProps } from '../model/TaskProps.types'
import { CommonInput } from '../common/CommonInput'
import { CommonButton } from '../common/CommonButton'
import { ListGroup, Form } from 'react-bootstrap'
import './Task.scss'

export const Task = ({ task, onEdit, onDelete, onComplete }: TaskProps) => {
    const [isEditing, setIsEditing] = useState<TaskProps['task']['completed']>(
        task.completed,
    )
    const [editedTask, setEditedTask] = useState<
        TaskProps['task']['description']
    >(task.description)
    const [validated, setValidated] = useState<boolean>(false)
    const [hideCheckbox, setHideCheckbox] = useState<boolean>(false)

    const handleSave = (e: React.FormEvent) => {
        e.preventDefault()

        if (editedTask === '') {
            setValidated(true)
        } else {
            onEdit(task.id, editedTask)
            setIsEditing(false)
        }
        setHideCheckbox(false)
    }
    const saveEditedTask = (e: React.ChangeEvent<HTMLInputElement>) => {
        setEditedTask(e.target.value)
    }
    const handleEditTask = () => {
        setIsEditing(true)
        setHideCheckbox(true)
    }

    const handleOnComplete = () => {
        onComplete(task.id)
    }

    const handleDelete = () => {
        onDelete(task.id)
    }

    return (
        <ListGroup className="task" horizontal={task.description.length <= 20}>
            {!hideCheckbox ? (
                <ListGroup.Item>
                    <CommonInput
                        type="checkbox"
                        className="input-checkbox"
                        data-testid="input-checkbox"
                        checked={task.completed}
                        handleChange={handleOnComplete}
                    />
                </ListGroup.Item>
            ) : null}
            {isEditing ? (
                <ListGroup.Item>
                    <Form
                        className="d-flex"
                        noValidate
                        validated={validated}
                        onSubmit={handleSave}
                    >
                        <CommonInput
                            className="input-text-edit"
                            data-testid="input-text-edit"
                            type="text"
                            value={editedTask}
                            checked={task.completed}
                            handleChange={saveEditedTask}
                        />
                        <CommonButton
                            text="Save"
                            variant="transparent"
                            type="submit"
                        />
                    </Form>
                </ListGroup.Item>
            ) : (
                <>
                    <ListGroup.Item className={'task-description'}>
                        {task.description}
                    </ListGroup.Item>
                    <ListGroup.Item>
                        <CommonButton
                            text="Edit"
                            variant="transparent"
                            type="button"
                            handleClick={handleEditTask}
                        />
                    </ListGroup.Item>
                </>
            )}
            <ListGroup.Item>
                <CommonButton
                    text="Delete"
                    variant="transparent"
                    type="button"
                    handleClick={handleDelete}
                />
            </ListGroup.Item>
        </ListGroup>
    )
}
