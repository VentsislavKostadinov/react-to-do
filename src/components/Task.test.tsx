import React from 'react'
import { Task } from './Task'
import { TaskProps } from '../model/TaskProps.types'
import { render } from '@testing-library/react'

describe('Task component', () => {
    const taskProps: TaskProps = {
        task: {
            id: 46456456456,
            description: 'text description',
            completed: false,
        },

        onEdit: jest.fn,
        onDelete: jest.fn,
        onComplete: jest.fn,
    }

    it('renders correctly', () => {
        const { asFragment } = render(<Task {...taskProps} />)
        expect(asFragment).toMatchSnapshot()
    })

    it('displayed correctly', () => {
        render(<Task {...taskProps} />)

        const container = document.querySelector('.container .col-lg-9')
        expect(container?.innerHTML).toContain('text description')
    })

})
