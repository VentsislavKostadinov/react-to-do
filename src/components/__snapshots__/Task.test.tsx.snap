// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task component renders correctly 1`] = `[Function]`;
